staging/src/k8s.io/apimachinery/pkg/runtime/serializer/versioning/versioning_test.go

```diff
@@ -380,3 +382,219 @@ func TestDirectCodecEncode(t *testing.T) {
+			encodes:    v1GVK.GroupVersion(),
+			errFunc: func(err error) bool {
+				return assert.Equal(t, runtime.NewNotRegisteredGVKErrForTarget("noxu-scheme", elseGVK, v1GVK.GroupVersion()), err)
+			},
+		},
+		{
+			name:       "encode with unstructured list's gvk regardless of its elements' gvk",
+			serializer: &mockSerializer{},
+			outObj:     elseUnstructuredDecodableList,
+			encodes:    elseGVK.GroupVersion(),
+		},
+		{
+			name:       "typer fail to recognize unstructured object gvk will fail the encoding",
+			serializer: &mockSerializer{},
+			outObj:     elseUnstructuredDecodable,
+			encodes:    v1GVK.GroupVersion(),
+			typer: &mockTyper{
+				err: fmt.Errorf("invalid obj gvk"),
+			},
+		},
+		{
+			name:       "encoding unstructured object without encode version will fallback to typer suggested version",
+			serializer: &mockSerializer{},
+			encodes:    v1GVK.GroupVersion(),
+			convertor: &checkConvertor{
+				obj:          v1UnstructuredDecodableList,
+				groupVersion: v1GVK.GroupVersion(),
+			},
+			outObj: elseUnstructuredDecodable,
+			typer: &mockTyper{
+				gvks: []schema.GroupVersionKind{v1GVK},
+			},
+		},
+	}
+	for _, testCase := range testCases {
+		codec := NewCodec(testCase.serializer, testCase.serializer, testCase.convertor, nil, testCase.typer, nil, testCase.encodes, nil, "noxu-scheme")
+		err := codec.Encode(testCase.outObj, ioutil.Discard)
+		if testCase.errFunc != nil {
+			if !testCase.errFunc(err) {
+				t.Errorf("%v: failed: %v", testCase.name, err)
+			}
+			return
+		}
+		assert.NoError(t, err)
+		assert.Equal(t, testCase.expectedObj, testCase.serializer.obj)
+	}
+}
+
+func TestDecodeUnstructured(t *testing.T) {
+	internalGVK := schema.GroupVersionKind{
+		Group:   "crispy",
+		Version: runtime.APIVersionInternal,
+		Kind:    "Noxu",
+	}
+	v1GVK := schema.GroupVersionKind{
+		Group:   "crispy",
+		Version: "v1",
+		Kind:    "Noxu",
+	}
+	v1UnstructuredDecodable := buildUnstructuredDecodable(v1GVK)
+	internalUnstructuredDecodable := buildUnstructuredDecodable(internalGVK)
+
+	testCases := []struct {
+		name       string
+		serializer runtime.Serializer
+		convertor  runtime.ObjectConvertor
+		decodes    runtime.GroupVersioner
+		defaultGVK *schema.GroupVersionKind
+		intoObj    runtime.Object
+
+		errFunc     func(error) bool
+		expectedGVK *schema.GroupVersionKind
+		expectedOut runtime.Object
+	}{
+		{
+			name:       "decode v1 unstructured into non-nil internal unstructured",
+			serializer: &mockSerializer{actual: &v1GVK, obj: v1UnstructuredDecodable},
+			convertor: &checkConvertor{
+				directConvert: true,
+				in:            v1UnstructuredDecodable,
+				obj:           internalUnstructuredDecodable,
+			},
+			decodes:     internalGVK.GroupVersion(),
+			intoObj:     internalUnstructuredDecodable,
+			expectedGVK: &v1GVK,
+			expectedOut: internalUnstructuredDecodable,
+		},
+		{
+			name:       "decode v1 unstructured into nil object with internal version",
```

Comment:
please add the same tests with v2 instead of internal. For CRDs we don't use internal types, but convert between two non-internal versions.